### DISCUSSION TITLE
Migrate to ES modules

### DISCUSSION THREAD

**Discussion by Lutymane** [haraka/Haraka#3421] [Ideas]
- [x] I already searched past issues
- [x] I already checked the [wiki/TODO](https://github.com/haraka/Haraka/wiki/TODO) page

### Is your feature request related to a problem? Please describe.

CommonJS is morally obsolete and it's hard to write some complex plugins and logic with this limitation (like reusing logic between SMTP and Web server written in modern code). Not talking about lack of top-level await in cjs

### Describe the solution you'd like

Migrate code base to ES modules

**Comment by msimerson** [comment]
Since moving from CommonJS to ESM is a moral imperative to you, we expect you'll be leading the effort. We look forward to all the PRs you will be submitting to convert our entire code base. I'd recommend starting at the very bottom and working your way up. [haraka-tld](https://github.com/Haraka/haraka-tld) would be a great place to start.

**Comment by msimerson** [comment]
This is a fun update: https://github.com/haraka/Haraka/issues/3564
